File: apps/mobile/test/macro_targets_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:logmyplate_mobile/src/models/macro_targets.dart';
import 'package:logmyplate_mobile/src/models/meal.dart';

/// Conformance suite for the shared Part A specification.
///
/// The same fixture is asserted by the TypeScript implementation in
/// packages/domain/src/macro-targets.test.ts. Pinning both to one set of vectors
/// is what makes it safe for the target screen to preview a calorie number and
/// macro bands locally while the API stays authoritative for what is saved.
///
/// This is not a theoretical safeguard. The screen kept the old flat -300/+250
/// goal offsets after the server moved to multiplicative factors, and showed
/// people a target roughly 200 kcal away from the one being stored.
///
/// Regenerate after a deliberate change to Part A:
///   pnpm --filter @logmyplate/domain build
///   node packages/domain/fixtures/generate-macro-target-vectors.mjs
void main() {
  final fixture = File(
    '../../packages/domain/fixtures/macro-target-vectors.json',
  );

  HealthSex sexFrom(String value) => switch (value) {
    'male' => HealthSex.male,
    'female' => HealthSex.female,
    _ => HealthSex.notSpecified,
  };

  ActivityLevel? activityFrom(String value) => switch (value) {
    'sedentary' => ActivityLevel.sedentary,
    'light' => ActivityLevel.light,
    'moderate' => ActivityLevel.moderate,
    'active' => ActivityLevel.active,
    // The app does not yet offer "extra_active"; the API accepts it, so these
    // vectors are skipped rather than mapped onto a level that means something
    // different.
    _ => null,
  };

  HealthGoal goalFrom(String value) => switch (value) {
    'lose_gently' => HealthGoal.loseGently,
    'gain_gently' => HealthGoal.gainGently,
    _ => HealthGoal.maintain,
  };

  test('shared vector fixture is present', () {
    expect(
      fixture.existsSync(),
      isTrue,
      reason:
          'Run: node packages/domain/fixtures/generate-macro-target-vectors.mjs',
    );
  });

  test('matches every shared macro target vector', () {
    final data = jsonDecode(fixture.readAsStringSync()) as Map<String, dynamic>;
    final vectors = data['vectors'] as List<dynamic>;
    expect(vectors, isNotEmpty);

    var checked = 0;
    for (final entry in vectors) {
      final vector = entry as Map<String, dynamic>;
      final input = vector['input'] as Map<String, dynamic>;
      final expected = vector['expected'] as Map<String, dynamic>;
      final name = vector['name'] as String;

      final activityLevel = activityFrom(input['activityLevel'] as String);
      if (activityLevel == null) continue;
      checked++;

      final actual = calculateMacroTargets(
        heightCm: (input['heightCm'] as num).toDouble(),
        weightKg: (input['weightKg'] as num).toDouble(),
        ageYears: (input['ageYears'] as num).toInt(),
        sex: sexFrom(input['sex'] as String),
        activityLevel: activityLevel,
        goal: goalFrom(input['goal'] as String),
        customMacroSplit: MacroSplit.fromJson(
          input['customMacroSplit'] as Map<String, dynamic>?,
        ),
      );

      expect(actual.bmi, (expected['bmi'] as num).toDouble(), reason: '$name bmi');
      expect(
        actual.targetDailyCalories,
        (expected['targetDailyCalories'] as num).toInt(),
        reason: '$name target calories',
      );
      expect(
        actual.bmrCalories,
        (expected['bmrCalories'] as num).toDouble(),
        reason: '$name bmr',
      );

      final centers = expected['centers'] as Map<String, dynamic>;
      expect(
        actual.centers.carbsPct,
        (centers['carbsPct'] as num).toDouble(),
        reason: '$name carbs centre',
      );
      expect(
        actual.centers.fatPct,
        (centers['fatPct'] as num).toDouble(),
        reason: '$name fat centre',
      );
      expect(
        actual.centers.proteinPct,
        (centers['proteinPct'] as num).toDouble(),
        reason: '$name protein centre',
      );

      final bands = expected['bands'] as Map<String, dynamic>;
      final carbs = bands['carbsPct'] as Map<String, dynamic>;
      expect(
        actual.carbsBand.min,
        (carbs['min'] as num).toDouble(),
        reason: '$name carbs band min',
      );
      expect(
        actual.carbsBand.max,
        (carbs['max'] as num).toDouble(),
        reason: '$name carbs band max',
      );

      expect(
        actual.customSplitApplied,
        expected['customSplitApplied'] as bool,
        reason: '$name custom split flag',
      );
    }

    // Guards against the activity filter silently skipping everything.
    expect(checked, greaterThan(40));
  });

  test('a custom split replaces the computed centres outright', () {
    final targets = calculateMacroTargets(
      heightCm: 175,
      weightKg: 70,
      ageYears: 30,
      sex: HealthSex.male,
      activityLevel: ActivityLevel.moderate,
      goal: HealthGoal.maintain,
      customMacroSplit: const MacroSplit(
        carbsPct: 30,
        fatPct: 25,
        proteinPct: 45,
      ),
    );

    // An explicit choice is a stronger signal of intent than an estimate, so it
    // replaces rather than blends.
    expect(targets.centers.carbsPct, 30);
    expect(targets.centers.proteinPct, 45);
    expect(targets.customSplitApplied, isTrue);
    expect(targets.tolerance, 5);
  });
}

File: packages/domain/src/macro-targets.test.ts
import { describe, expect, it } from "vitest";
import {
  calculateBmr,
  calculateMacroTargets,
  type MacroActivityLevel,
  type MacroGoal,
  type MacroTargetInput,
} from "./macro-targets.js";

/** The worked-example user: female, 28, 65kg, 165cm, lightly active, muscle gain. */
const workedExample: MacroTargetInput = {
  heightCm: 165,
  weightKg: 65,
  ageYears: 28,
  sex: "female",
  activityLevel: "light",
  goal: "gain_gently",
};

describe("calculateBmr", () => {
  it("follows Mifflin-St Jeor for female", () => {
    // 10(65) + 6.25(165) - 5(28) - 161
    expect(calculateBmr(workedExample)).toBeCloseTo(1380.25, 2);
  });

  it("follows Mifflin-St Jeor for male", () => {
    expect(calculateBmr({ ...workedExample, sex: "male" })).toBeCloseTo(1546.25, 2);
  });

  it("averages the two when sex is not specified", () => {
    const female = calculateBmr({ ...workedExample, sex: "female" });
    const male = calculateBmr({ ...workedExample, sex: "male" });
    expect(calculateBmr({ ...workedExample, sex: "not_specified" })).toBeCloseTo(
      (female + male) / 2,
      2,
    );
  });
});

describe("calculateMacroTargets", () => {
  it("reproduces the worked example's BMI, centres, tolerance and bands", () => {
    const result = calculateMacroTargets(workedExample);

    // These all agree with meal-health-score-worked-example.md.
    expect(result.bmi).toBe(23.9);
    expect(result.centers).toEqual({ carbsPct: 40, fatPct: 25, proteinPct: 35 });
    expect(result.tolerance).toBe(5);
    expect(result.bands).toEqual({
      carbsPct: { min: 35, max: 45 },
      fatPct: { min: 20, max: 30 },
      proteinPct: { min: 30, max: 40 },
    });
  });

  it("computes calories from the formula, not the worked example's arithmetic", () => {
    const result = calculateMacroTargets(workedExample);

    // The worked example states BMR 1,346 / TDEE 1,851 / target 2,036. That BMR
    // does not follow Mifflin-St Jeor (it matches a height of ~159.5cm) and the
    // error cascades. Part A as written gives these values.
    expect(result.bmrCalories).toBeCloseTo(1380.3, 1);
    expect(result.tdeeCalories).toBeCloseTo(1897.8, 1);
    expect(result.targetDailyCalories).toBe(2088);
  });

  describe("A3 activity factors", () => {
    const factors: Array<[MacroActivityLevel, number]> = [
      ["sedentary", 1.2],
      ["light", 1.375],
      ["moderate", 1.55],
      ["active", 1.725],
      ["extra_active", 1.9],
    ];

    for (const [level, factor] of factors) {
      it(`scales TDEE by ${factor} for ${level}`, () => {
        const result = calculateMacroTargets({ ...workedExample, activityLevel: level });
        expect(result.tdeeCalories).toBeCloseTo(1380.25 * factor, 1);
      });
    }
  });

  describe("A4 goal factors are multiplicative", () => {
    const factors: Array<[MacroGoal, number]> = [
      ["lose_gently", 0.8],
      ["maintain", 1.0],
      ["gain_gently", 1.1],
    ];

    for (const [goal, factor] of factors) {
      it(`applies x${factor} for ${goal}`, () => {
        const result = calculateMacroTargets({ ...workedExample, goal });
        expect(result.targetDailyCalories).toBe(Math.round(1380.25 * 1.375 * factor));
      });
    }
  });

  describe("A5 macro centres by goal x activity", () => {
    // Every cell of the lookup table, so a typo cannot slip through.
    const cases: Array<[MacroGoal, MacroActivityLevel, number, number, number]> = [
      ["lose_gently", "sedentary", 30, 30, 40],
      ["lose_gently", "light", 30, 30, 40],
      ["lose_gently", "moderate", 35, 25, 40],
      ["lose_gently", "active", 35, 25, 40],
      ["lose_gently", "extra_active", 35, 25, 40],
      ["gain_gently", "sedentary", 40, 25, 35],
      ["gain_gently", "light", 40, 25, 35],
      ["gain_gently", "moderate", 45, 25, 30],
      ["gain_gently", "active", 45, 25, 30],
      ["gain_gently", "extra_active", 45, 25, 30],
      ["maintain", "sedentary", 45, 30, 25],
      ["maintain", "light", 45, 30, 25],
      ["maintain", "moderate", 50, 25, 25],
      ["maintain", "active", 50, 25, 25],
      ["maintain", "extra_active", 50, 25, 25],
    ];

    for (const [goal, activityLevel, carbsPct, fatPct, proteinPct] of cases) {
      it(`${goal} + ${activityLevel} -> ${carbsPct}/${fatPct}/${proteinPct}`, () => {
        const result = calculateMacroTargets({ ...workedExample, goal, activityLevel });
        expect(result.centers).toEqual({ carbsPct, fatPct, proteinPct });
      });
    }

    it("sums centres to 100 in every cell", () => {
      for (const [goal, activityLevel] of cases) {
        const { centers } = calculateMacroTargets({ ...workedExample, goal, activityLevel });
        expect(centers.carbsPct + centers.fatPct + centers.proteinPct).toBe(100);
      }
    });
  });

  describe("A6 BMI nudge", () => {
    it("shifts carbs to protein when BMI is 30 or above", () => {
      // 95kg at 165cm -> BMI 34.9
      const result = calculateMacroTargets({ ...workedExample, weightKg: 95 });

      expect(result.bmi).toBeGreaterThanOrEqual(30);
      expect(result.centers).toEqual({ carbsPct: 35, fatPct: 25, proteinPct: 40 });
    });

    it("shifts toward carbs and fat when BMI is under 18.5", () => {
      // 45kg at 165cm -> BMI 16.5
      const result = calculateMacroTargets({ ...workedExample, weightKg: 45 });

      expect(result.bmi).toBeLessThan(18.5);
      expect(result.centers).toEqual({ carbsPct: 45, fatPct: 28, proteinPct: 32 });
    });

    it("leaves a normal BMI untouched", () => {
      const result = calculateMacroTargets(workedExample);
      expect(result.centers).toEqual({ carbsPct: 40, fatPct: 25, proteinPct: 35 });
    });

    it("is additive, so goal and activity still drive the shape", () => {
      // An obese maintenance user keeps maintenance's fat centre; only the
      // carb/protein pair moves.
      const result = calculateMacroTargets({
        ...workedExample,
        weightKg: 95,
        goal: "maintain",
        activityLevel: "sedentary",
      });
      expect(result.centers).toEqual({ carbsPct: 40, fatPct: 30, proteinPct: 30 });
    });
  });

  describe("A7 tolerance", () => {
    it("is tighter for goal-driven users", () => {
      expect(calculateMacroTargets({ ...workedExample, goal: "lose_gently" }).tolerance).toBe(5);
      expect(calculateMacroTargets({ ...workedExample, goal: "gain_gently" }).tolerance).toBe(5);
    });

    it("is looser for maintenance", () => {
      expect(calculateMacroTargets({ ...workedExample, goal: "maintain" }).tolerance).toBe(8);
    });
  });

  describe("A9 manual override", () => {
    it("replaces the computed centres outright", () => {
      const result = calculateMacroTargets({
        ...workedExample,
        customMacroSplit: { carbsPct: 50, fatPct: 20, proteinPct: 30 },
      });

      expect(result.customSplitApplied).toBe(true);
      expect(result.centers).toEqual({ carbsPct: 50, fatPct: 20, proteinPct: 30 });
      expect(result.tolerance).toBe(5);
      expect(result.bands.carbsPct).toEqual({ min: 45, max: 55 });
    });

    it("ignores the BMI nudge, since an explicit choice outranks it", () => {
      const result = calculateMacroTargets({
        ...workedExample,
        weightKg: 95,
        customMacroSplit: { carbsPct: 50, fatPct: 20, proteinPct: 30 },
      });

      expect(result.centers.carbsPct).toBe(50);
    });

    it("reports no override when none was given", () => {
      expect(calculateMacroTargets(workedExample).customSplitApplied).toBe(false);
    });
  });

  it("is deterministic", () => {
    expect(calculateMacroTargets(workedExample)).toEqual(calculateMacroTargets(workedExample));
  });
});
